### PR TITLE
[linky] Adressing issue #11642

### PR DESCRIPTION
**Pull request by clinique**
Apparently all the mess came from default write buffer size of Jetty client. 
Corrected it by adding a parameter at binding configuration level.
Maybe this was generated during an update of jetty lib in the core.
I also corrected daily data that does not seem to accept J+1 anymore and three SAT findings.

Signed-off-by: clinique <gael@lhopital.org>

### CODE REVIEW

**Comment by clinique** [comment]
I made some testings with this requestBufferSize, starting from 32k then 16k then 8k. It broke at 4k, so I ended at 6000.

**Comment by lolodomo** [comment]
The parameter is ultra technical. Is it really necessary ? Why not just a constant in the code with an appropriate value ?

I am going to test your fix.
Thank you a lot for fixing that, I am impressed how fast you found the issue.

**Comment by lolodomo** [comment]
Your binding version does not start:
```
13:10:48.475 [WARN ] [org.apache.felix.fileinstall         ] - Error while starting bundle: file:/home/pi/openhab/addons/org.openhab.binding.linky-3.3.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.linky [276]
  Unresolved requirement: Import-Package: org.jsoup; version="[1.14.0,2.0.0)"

        at org.eclipse.osgi.container.Module.start(Module.java:463) ~[org.eclipse.osgi-3.16.300.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:445) ~[org.eclipse.osgi-3.16.300.jar:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [bundleFile:3.7.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [bundleFile:3.7.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520) [bundleFile:3.7.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [bundleFile:3.7.4]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [bundleFile:3.7.4]
```

**Comment by lolodomo** [comment]
The problem is the dependency. I can't use a simple jar in this case. I probably need a kar file including the jar and the dependency. Do you know what is the maven command to build this kar file ?

**Comment by lolodomo** [comment]
I know a workaround, starting with the official binding and then simply updating the bundle. I try...

**Comment by clinique** [comment]
> The parameter is ultra technical. Is it really necessary ? Why not just a constant in the code with an appropriate value ?
> 
> I am going to test your fix. Thank you a lot for fixing that, I am impressed how fast you found the issue.

This was my first approach and I thought that due to the number of ways this binding explodes for external reasons, having this parameter easily changeable could be usefull.

**Comment by clinique** [comment]
> The problem is the dependency. I can't use a simple jar in this case. I probably need a kar file including the jar and the dependency. Do you know what is the maven command to build this kar file ?

I had the same problem, in testing I did change the 'provided' by 'compile' in the pom.xml file.

**Comment by lolodomo** [comment]
My solution worked but I was forced to then restart OH. If not, the binding was showing no more channels.

Regarding the parameter, I did not realize it was on the binding itself, I was searching it on my thing in MainUI. I am still not convinced that such a parameter is a good idea for a final user. In case you really want to keep it, it should be documented (in README).

I have now the binding retrieving my data. Good. But I remember that sometimes it was initially working and then failing at the next retrieval of data. So let's see how it will behave tomorrow.

**Comment by clinique** [comment]
> Regarding the parameter, I did not realize it was on the binding itself, I was searching it on my thing in MainUI. I am still not convinced that such a parameter is a good idea for a final user. In case you really want to keep it, it should be documented (in README).
> 

I did it. Maybe you would expect more informations ?

File: bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:111]: Are you sure it is not necessary to minus 2 days to be be certain that the previous day will be fully covered ?
I don't know if time is considered or only date.

File: bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky.properties
Comment (lolodomo) [inline bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky.properties:6]: Maybe you could make description here and in the XML file in sync.

File: bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky_fr.properties
Comment (lolodomo) [inline bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky_fr.properties:4]: Please do not update the French translation in this PR.
By the way, you apparently even introduce a problem, maybe due to an invalid file encoding:
image

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
Looks like the problem is not resolved. When I linked the power channel, the authentication is failing, I got this:
```
14:12:31.033 [DEBUG] [g.linky.internal.handler.LinkyHandler] - Refreshing channel daily#power
14:12:31.037 [DEBUG] [g.linky.internal.api.ExpiringDayCache] - getValue from cache "power cache" is requiring a fresh value
14:12:31.040 [DEBUG] [g.linky.internal.handler.LinkyHandler] - getPowerData from 2022-04-01 to 2022-04-02
14:12:31.043 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Starting login process for user : xxxxx
14:12:31.046 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Step 1 : getting authentification
14:12:31.404 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Reception request SAML
14:12:31.412 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Step 2 : send SSO SAMLRequest
14:12:31.677 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Get the location and the ReqID
14:12:31.680 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - Step 3 : auth1 - retrieve the template, thanks to cookie internalAuthId, user is already set
14:12:31.937 [DEBUG] [ding.linky.internal.api.EnedisHttpApi] - auth1 - invalid template for auth data: {"authId":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","template":"","stage":"EnedisLogin2","header":"Tout Enedis en un seul compte","callbacks":[{"type":"NameCallback","output":[{"name":"prompt","value":"E-mail"}],"input":[{"name":"IDToken1","value":""}]},{"type":"CaptchaCallback","output":[{"name":"version","value":"V2"},{"name":"prompt","value":"yyyyyyyyyyyyyy"}],"input":[{"name":"IDToken2","value":""}]},{"type":"ConfirmationCallback","output":[{"name":"prompt","value":""},{"name":"messageType","value":0},{"name":"options","value":["Suivant"]},{"name":"optionType","value":-1},{"name":"defaultOption","value":0}],"input":[{"name":"IDToken3","value":0}]},{"type":"TextOutputCallback","output":[{"name":"message","value":"Espace client ENEDIS"},{"name":"messageType","value":"0"}]},{"type":"TextOutputCallback","output":[{"name":"message","value":"GdcUrl"},{"name":"messageType","value":"0"}]}]}
14:12:31.941 [DEBUG] [g.linky.internal.handler.LinkyHandler] - Exception when getting power data: Authentication error, the authentication_cookie is probably wrong
org.openhab.binding.linky.internal.LinkyException: Authentication error, the authentication_cookie is probably wrong
        at org.openhab.binding.linky.internal.api.EnedisHttpApi.initialize(EnedisHttpApi.java:124) ~[?:?]
        at org.openhab.binding.linky.internal.api.EnedisHttpApi.getPowerData(EnedisHttpApi.java:290) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.getPowerData(LinkyHandler.java:384) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.lambda$1(LinkyHandler.java:112) ~[?:?]
        at org.openhab.binding.linky.internal.api.ExpiringDayCache.refreshValue(ExpiringDayCache.java:86) ~[?:?]
        at org.openhab.binding.linky.internal.api.ExpiringDayCache.getValue(ExpiringDayCache.java:73) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.updatePowerData(LinkyHandler.java:206) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.handleCommand(LinkyHandler.java:444) ~[?:?]
        at org.openhab.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:180) ~[?:?]
        at org.openhab.core.thing.internal.ChannelLinkNotifier.lambda$3(ChannelLinkNotifier.java:72) ~[?:?]
        at org.openhab.core.thing.internal.ChannelLinkNotifier.call(ChannelLinkNotifier.java:96) ~[?:?]
        at org.openhab.core.thing.internal.ChannelLinkNotifier.added(ChannelLinkNotifier.java:72) ~[?:?]
        at org.openhab.core.thing.internal.ChannelLinkNotifier.added(ChannelLinkNotifier.java:1) ~[?:?]
        at org.openhab.core.common.registry.AbstractRegistry.notifyListeners(AbstractRegistry.java:367) ~[?:?]
        at org.openhab.core.common.registry.AbstractRegistry.notifyListenersAboutAddedElement(AbstractRegistry.java:400) ~[?:?]
        at org.openhab.core.thing.link.ItemChannelLinkRegistry.notifyListenersAboutAddedElement(ItemChannelLinkRegistry.java:138) ~[?:?]
        at org.openhab.core.thing.link.ItemChannelLinkRegistry.notifyListenersAboutAddedElement(ItemChannelLinkRegistry.java:1) ~[?:?]
        at org.openhab.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:175) ~[?:?]
        at org.openhab.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:1) ~[?:?]
        at org.openhab.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:60) ~[?:?]
        at org.openhab.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:79) ~[?:?]
        at org.openhab.core.common.registry.AbstractProvider.notifyListenersAboutAddedElement(AbstractProvider.java:83) ~[?:?]
        at org.openhab.core.common.registry.AbstractManagedProvider.add(AbstractManagedProvider.java:67) ~[?:?]
        at org.openhab.core.common.registry.AbstractRegistry.add(AbstractRegistry.java:346) ~[?:?]
        at org.openhab.core.io.rest.core.internal.link.ItemChannelLinkResource.link(ItemChannelLinkResource.java:176) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) ~[bundleFile:3.4.5]
        at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) ~[bundleFile:3.4.5]
        at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201) ~[bundleFile:3.4.5]
        at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104) ~[bundleFile:3.4.5]
        at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) ~[bundleFile:3.4.5]
        at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) ~[bundleFile:3.4.5]
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298) ~[bundleFile:3.4.5]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPut(AbstractHTTPServlet.java:234) ~[bundleFile:3.4.5]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:710) ~[bundleFile:3.1.0]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273) ~[bundleFile:3.4.5]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:294) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:82) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.Server.handle(Server.java:516) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:386) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [bundleFile:9.4.43.v20210629]
        at java.lang.Thread.run(Thread.java:834) [?:?]
14:12:32.169 [DEBUG] [g.linky.internal.api.ExpiringDayCache] - calcNextExpiresAt result = 2022-04-03T01:00:00
14:12:32.170 [INFO ] [hab.event.ThingStatusInfoChangedEvent] - Thing 'linky:linky:local' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Authentication error, the authentication_cookie is probably wrong
14:12:32.175 [DEBUG] [g.linky.internal.handler.LinkyHandler] - Update channel daily#power with NaN
```

**Comment by lolodomo** [comment]
In fact, reading again my issue, my feeling now is that the problem is still exactly the same...

File: bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
Comment (clinique) [reply bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:111]: The request is only date based. When I made the test, keeping minusDays(2), I landed on the 31st of March.

File: bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky_fr.properties
Comment (clinique) [reply bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky_fr.properties:4]: Oh, shit ! I corrected all the file thinking I introduced an incorrect encoding. I'll revert it to original version and we'll take care of it in crowdin.

File: bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky.properties
Comment (clinique) [reply bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky.properties:6]: Done

**Comment by clinique** [comment]
> In fact, reading again my issue, my feeling now is that the problem is still exactly the same...

Would this mean that it works fine on first run and fails on refresh ?

**Comment by lolodomo** [comment]
> Would this mean that it works fine on first run and fails on refresh ?

Not sure it works systematically at first run but what is sure is that it fails systematically at time of refresh.
But at least it can (partially) works at first run.

**Comment by lolodomo** [comment]
I just restarted the bundle and it works well, all the API calls succeeded.
But I am sure that if I trigger a new API call in few minutes, for example by unlinking/linking a channel, it will fail again.
Will try it in few minutes.

**Comment by lolodomo** [comment]
Confirmed. I restarted without the power peak channel linked. No problem. Few minutes later, I link the power peak channel which trigger a API call to get the value and it failed.
So yes, it looks like the initial calls at binding initialization are ok but later any refresh is then failing.

File: bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
Comment (lolodomo) [reply bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:111]: Ok

**Comment by clinique** [comment]
> Confirmed. I restarted without the power peak channel linked. No problem, API calls succeeded. Few minutes later, I link the power peak channel which triggered a API call to get the value and it failed. So yes, it looks like the initial calls at binding initialization are ok but later any refresh is then failing.

I unfortunately do not reproduce. Everything seems to work fine on my side.
I unlinked & removed peak channel. Restarted OH. Values correctly populated. Added a link to peak that got populated correctly.
I am going to investigate to understand what could be left between two updates but I'm like a blind in a dark room.

**Comment by clinique** [comment]
> Confirmed. I restarted without the power peak channel linked. No problem, API calls succeeded. Few minutes later, I link the power peak channel which triggered a API call to get the value and it failed. So yes, it looks like the initial calls at binding initialization are ok but later any refresh is then failing.

Maybe have I found something : the internalAuthId is added in the coockieSet at EnedisHttpApi creation while it should be done at initialization because it is removed on disconnection and initialization maybe called without recreation of the object.

**Comment by lolodomo** [comment]
Bravo, I have the feeling that it works well now.

**Comment by lolodomo** [comment]
I ask again my question: do you really think there will be in the future the need to adjust this buffer size ?
Why not a constant and you take a good margin ?

**Comment by lolodomo** [comment]
I will see how it behaves tomorrow (when it is time for a refresh).

**Comment by clinique** [comment]
> I ask again my question: do you really think there will be in the future the need to adjust this buffer size ? Why not a constant in the code and you take a good margin ?

Ok, once I have your feed-back tomorrow morning, no further modification needed, I'll revert this to a constant.

**Comment by lolodomo** [comment]
@clinique : works as expected this morning. Thing is still ONLINE. 

Waiting for your last change before I merge.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you so much

**Comment by clinique** [comment]
> LGTM, thank you so much

...until something else changes on Enedis website... :)

**Pull request merged**
